src/vs/platform/quickinput/browser/quickPickPin.ts

```diff
@@ -0,0 +1,107 @@
+
+import { Codicon } from 'vs/base/common/codicons';
+import { localize } from 'vs/nls';
+import { ServicesAccessor } from 'vs/platform/instantiation/common/instantiation';
+import { IQuickPick, IQuickPickItem, QuickPickItem } from 'vs/platform/quickinput/common/quickInput';
+import { IStorageService, StorageScope, StorageTarget } from 'vs/platform/storage/common/storage';
+import { ThemeIcon } from 'vs/platform/theme/common/themeService';
+
+const pinButtonClass = ThemeIcon.asClassName(Codicon.pin);
+const pinnedButtonClass = ThemeIcon.asClassName(Codicon.pinned);
+const buttonClasses = [pinButtonClass, pinnedButtonClass];
+/**
+ * Initially, adds pin buttons to all @param quickPick items.
+ * When pinned, a copy of the item will be moved to the end of the pinned list and any duplicate within the pinned list will
+ * be removed if @param filterDupliates has been provided. Pin and pinned button events trigger updates to the underlying storage.
+ * Shows the quickpick once formatted.
+ */
+export async function showWithPinnedItems(accessor: ServicesAccessor, storageKey: string, quickPick: IQuickPick<IQuickPickItem>, filterDuplicates?: boolean): Promise<void> {
+	const storageService = accessor.get(IStorageService);
+	quickPick.onDidTriggerItemButton(async buttonEvent => {
+		const expectedButton = buttonEvent.button.iconClass && buttonClasses.includes(buttonEvent.button.iconClass);
+		if (expectedButton) {
+			quickPick.items = await _formatPinnedItems(storageKey, quickPick, storageService, buttonEvent.item, filterDuplicates);
+		}
+	});
+	quickPick.onDidChangeValue(async value => {
+		// don't show pinned items in the search results
+		quickPick.items = value ? quickPick.items.filter(i => i.type !== 'separator' && !i.buttons?.find(b => b.iconClass === pinnedButtonClass)) : quickPick.items;
+	});
+	quickPick.items = await _formatPinnedItems(storageKey, quickPick, storageService, undefined, filterDuplicates);
+	await quickPick.show();
+}
+
+function _formatPinnedItems(storageKey: string, quickPick: IQuickPick<IQuickPickItem>, storageService: IStorageService, changedItem?: IQuickPickItem, filterDuplicates?: boolean): QuickPickItem[] {
+	const formattedItems: QuickPickItem[] = [];
+	let pinnedItems;
+	if (changedItem) {
+		pinnedItems = updatePinnedItems(storageKey, changedItem, storageService);
+	} else {
+		pinnedItems = getPinnedItems(storageKey, storageService);
+	}
+	if (pinnedItems.length) {
+		formattedItems.push({ type: 'separator', label: localize("terminal.commands.pinned", 'Pinned') });
+	}
+	const pinnedIds = new Set();
+	for (const itemToFind of pinnedItems) {
+		const itemToPin = quickPick.items.find(item => itemsMatch(item, itemToFind));
+		if (itemToPin) {
+			const pinnedItemId = getItemIdentifier(itemToPin);
+			const pinnedItem: IQuickPickItem = Object.assign({} as IQuickPickItem, itemToPin);
+			if (!filterDuplicates || !pinnedIds.has(pinnedItemId)) {
+				pinnedItem.id = pinnedItemId;
+				pinnedIds.add(pinnedItem.id);
```

Comment:
```suggestion
				pinnedIds.add(pinnedItemId);
```
I think this only needs to be this? Since `pinnedItem.id` will either already be set to a value or we just rely on the `getItemIdentifier` to get the "placeholder id" for us.